src/Npgsql/HardCancellationBlock.cs

```diff
@@ -0,0 +1,33 @@
+﻿using System;
+using System.Threading;
+
+namespace Npgsql
+{
+    internal class HardCancellationBlock : IDisposable
```

Comment:
FYI, we prefer not to write default modifiers like `internal` for classes and seal everything which isn't designed for inheritance. For that case `struct` is better to avoid allocations. An object of that type will be incapsulated into a state machine directly at call sites.

Reply:
And is it really OK to have a mutable struct?

Reply:
As an option (to have an immutable struct and allow nested blocks), we can just return the previous state on dispose.

Reply:
`ManualResetValueTaskSourceCore<T>` is an example of a mutable struct used by state machines producing value tasks.

Reply:
It's always a good idea to avoid mutable structs when possible :) But in some cases they're needed.